osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs

```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
                 length * MathF.Sin(angle)
             );
         }
+
+        public static bool IsHitObjectOnBeat(OsuBeatmap beatmap, OsuHitObject hitObject, bool downbeatsOnly = false)
+        {
+            var timingPoints = beatmap.ControlPointInfo.TimingPoints;
+            var currentTimingPoint = timingPoints.Reverse().FirstOrDefault(p => p.Time <= hitObject.StartTime);
+
+            if (currentTimingPoint == null)
+                return false;
+
+            double timeSinceTimingPoint = hitObject.StartTime - currentTimingPoint.Time;
+
+            double length = downbeatsOnly
+                ? currentTimingPoint.BeatLength * currentTimingPoint.TimeSignature.Numerator
+                : currentTimingPoint.BeatLength;
+
+            return (timeSinceTimingPoint + 1) % length < 2;
```

Comment:
- Why `timeSinceTimingPoint + 1`?
- I would strongly advise against using the `%` operator on floats. It can return results that are just not correct if unlucky with floating point representation issues. Recommend using rounded division instead.

Reply:
This already contains a tolerance of +-1. With infinite precision, this would already work with returning `timeSinceTimingPoint % length == 0`, but by adding 1 and using `< 2` instead of `== 0` I made sure that no floating point issues occur. I also tested this method and I didn't find a single hit object on a (down-)beat that wasn't correctly identified by this method.